Merge mozilla-browsertime changes into sitespeedio-browsertime

gmierz opened the pull request:
This PR is adding the commits from mozilla-browsertime into the sitespeedio-browsertime repo (upstreaming the changes).

The following files had merge conflicts so they are of most interest in the reviews:
```
U	lib/android/index.js
U	lib/chrome/webdriver/index.js
U	lib/core/defaultPageCompleteCheck.js
U	lib/core/engine/collector.js
U	lib/core/engine/command/javaScript.js
U	lib/core/engine/command/measure.js
U	lib/core/engine/index.js
U	lib/core/engine/iteration.js
U	lib/core/seleniumRunner.js
U	lib/firefox/webdriver/firefoxDelegate.js
U	lib/firefox/webdriver/firefoxPreferences.js
U	lib/firefox/webdriver/index.js
U	lib/screenshot/index.js
U	lib/support/cli.js
U	lib/support/har/index.js
U	lib/video/screenRecording/firefox/firefoxWindowRecorder.js
U	lib/video/screenRecording/recorder.js
U	lib/video/video.js
U	npm-shrinkwrap.json
U	package.json
U	vendor/visualmetrics.py
```

gmierz commented:
TODO: Remove this commit from the upstreamed changes: https://github.com/sitespeedio/browsertime/pull/1028/commits/ec43310dedc2e9b2e49574edb5043f483dcf34d3

soulgalore commented:
Wow this will be awesome! I normally don't merge so large PRs though, but understand that's almost the only way so lets go! I think this will need to go as a new major release, but thats no problem.

I need to setup a server to continuously run some tests (as we do with sitespeed.io) before we merge, I'll do that tomorrow. Wanna have it running some time before we merge so we easily can find things (there are couple of issues from the latest merges #1027 #1013 #983 , so it could be that we find more things when we merge). 

Let me know when you think is ready for me to look into the PR!

soulgalore commented:
I've setup an instance but it seems the autobuild on Docker doesn't include your PR; maybe only branches in the project. My idea was to just switch to an auto build there, but I guess we could just build this branch on the server.

Can you @gmierz make it go through the Travis tests `npm run lint:fix` + there are some things breaking, then I can make it run there + go through the commits, I think I have a couple of questions :)

gmierz commented:
Sounds good - I'm happy to find a way to split this up if needed.

@soulgalore, the geckoprofiler changes have the `fs` package as a dependency but I see that it is a restricted module. Would you know of an alternative we could use here? It's being used here right now: https://github.com/sitespeedio/browsertime/blob/5f97b5c832bca0933a204372d58a54c1211f52c1/lib/core/engine/iteration.js#L237

soulgalore commented:
No lets go ahead with this @gmierz thats ok :)

My plan is to review tomorrow and then hopefully merge the coming week so that is fixed before Christmas.

> is a restricted module.

No you could use that or?

gmierz commented:
Ok, I ask because it's one of the linting failures: https://travis-ci.org/sitespeedio/browsertime/builds/624336797#L768

I would have to remove `fs` from here to use it - if that's ok with you then I'll make the change: https://github.com/sitespeedio/browsertime/blob/1ef54896a708949ed3d858784de71a3137b6f626/lib/core/.eslintrc.json#L3

soulgalore commented:
> I would have to remove fs from here to use it - if that's ok with you then I'll make the change:

Ah I see. I think the reason is that we used that to make sure we moved all file handling to the storage manager (my guess is that from the beginning it was all over the place). Either we can just add a utility method or as you said just remove that rule, do what you think is best :)

gmierz commented:
@soulgalore a small update on progress.
1. Added a method to StorageManager to be able to read files (for the geckoprofiler).
2. Added a fix for the geckodriver not being found by the selenium-webdriver.
3. Added a fix to check if an extra field exists before it is used (we added those changes).

At the moment I'm looking into an issue where the test scripts are being recognized as problematic and are hitting this error: https://github.com/sitespeedio/browsertime/pull/1028/files#diff-598e2038200e98260b612f9a4fc397d7R560

I think I will have to change the test scripts being used but maybe there's another way.

gmierz commented:
I found a better solution for that last problem. :)

soulgalore commented:
I looked into the problem and the issue is that we do multiple calls to video.stop. That part was moved on the main branch to happen before we collect all video metric I guess that caused the merge to add it twice. I didn't have push rights to your repo so this is the change that is needed:

Replace https://github.com/gmierz/browsertime/blob/mozbtimesync/lib/core/engine/command/measure.js#L270-L273

with:
https://github.com/gmierz/browsertime/blob/mozbtimesync/lib/core/engine/command/measure.js#L315-L325

soulgalore commented on lib/core/defaultPageCompleteCheck.js:
```diff
@@ -2,9 +2,9 @@
-    try { 
-        const end = window.performance.timing.loadEventEnd;
-        return (end > 0) && (Date.now() > end + waitTime);
+    try {
```
So we skip the extra wait time now, is there a reason for that? That is simple wait for the user to add more wait instead of changing the more complicated pageCompleteCheck

soulgalore commented on lib/core/defaultPageCompleteCheck.js:
```diff
@@ -2,9 +2,9 @@
-        return (end > 0) && (Date.now() > end + waitTime);
+    try {
+        var end = window.performance.timing.loadEventEnd;
+        return end > 0;
```
We skip the use of the extra waitTime, any reason for that? That's an way for the user to make the test test longer without using the more complicated pageCompleteCheck.

gmierz replied:
The reason we removed it here is because when we playback a recorded webpage (through mitmproxy) on Chrome, this code hangs. We moved this wait to here - I can change it back or we can try to find another solution to this: https://github.com/sitespeedio/browsertime/pull/1028/files/88bd5c79e52e2fb4b1ec2263bb1510666fd41ca2#diff-598e2038200e98260b612f9a4fc397d7R178

soulgalore replied:
Ah I see, we do the same for WebPageReplay since the date object is static: https://github.com/sitespeedio/browsertime/blob/master/docker/scripts/start.sh#L91 

Push this as is and the I can change the code so we use performance.now instead (or do you also override that?). Since we plan to release this as a major, it would be nice to fix so it works out of the box both and without a reply proxy, that will make things simpler.

gmierz replied:
Sounds good, and I don't think we override performance.now anywhere.

soulgalore commented on lib/core/seleniumRunner.js:
```diff
@@ -425,9 +427,18 @@ class SeleniumRunner {
-          cause: e
-        });
+        // Work around GeckoView Bug 1298921: https://bugzilla.mozilla.org/show_bug.cgi?id=1298921.
+        const bug_1298921 =
```
Haha I love practical programming :D

soulgalore commented on lib/firefox/webdriver/firefoxPreferences.js:
```diff
@@ -12,35 +12,9 @@ module.exports = {
   // Use hidden TTCP
   'dom.performance.time_to_contentful_paint.enabled': true,
```
What about when we do this release switch to use your defaults instead and release it as a major release? Right now these are` --firefox.mozillaProPreferences` but I think we can just make those default in later PR or do you see a problem with that?

gmierz replied:
I'll undo these changes and we can do that in a later PR.

soulgalore commented on vendor/visualmetrics.py:
```diff
@@ -1521,11 +1544,18 @@ def calculate_perceptual_speed_index(progress, directory):
 
 
 def calculate_hero_time(progress, directory, hero, viewport):
-  try:
```
The try/except needs to be included in the PR. We wrap the hero times because it fails on Arabic Wikipedia making all visual metrics fail. Think try/except is needed both hereto and Contentful speed index

gmierz replied:
Ok, I've added this back. The contentful speed index function has it as well.

gmierz commented:
Thanks @soulgalore for pointing me to the issue there! I'll update the patch with those changes.

I'm currently wrangling with some issues running this merge in production. For some reason the preferences we provide in a browser profile are being ignored and because of this we don't get many metrics (this test run only has the "loadtime" metric): https://treeherder.mozilla.org/#/jobs?repo=try&tier=1%2C2%2C3&revision=7cc129553a3b180855dba070a69291a31095982b&selectedJob=281196445

The reason I say that the preferences are being ignored is because when I add the preferences we need into the default Firefox Browsertime preferences, we get the metrics even though they've already been provided within the profile.

soulgalore commented:
@gmierz I think you can move https://github.com/sitespeedio/browsertime/blob/master/lib/firefox/webdriver/firefoxPreferencesMozillaPerf.js them to the default settings in your PR just to make it easy, then it "should" just work.

gmierz commented:
@soulgalore I ended up using the new `--firefox.mozillaProPreferences` for that and it's working now. :)

Here's the latest test run: https://treeherder.mozilla.org/#/jobs?repo=try&tier=1%2C2%2C3&revision=f5cd6290da0161746dac6a6d8561df664c31d633&selectedJob=281197852

soulgalore commented:
Lets fix what's missing in master :)
